File: 2015/day07.cpp
// https://adventofcode.com/2015/day/7

#include "../common/common.h"

#include <algorithm>
#include <array>
#include <cstdlib>
#include <iostream>
#include <print>
#include <ranges>
#include <span>
#include <string>
#include <string_view>
#include <tuple>

// A signal can be a wire ID or an an actual value
// We unify them by storing values as 16-bit numbers,
// and the wire ID as numbers that are higher than all signal values
using signal_t = std::uint32_t;
using wire_id_t = signal_t;
using strict_signal_t = std::uint16_t;
constexpr const wire_id_t first_wire_id = 1u << 16;
constexpr bool is_value(signal_t signal) { return signal < first_wire_id; }
constexpr size_t input_id(wire_id_t wire_id) {
  AOC_ASSERT(wire_id >= first_wire_id, "Invalid wire ID");
  return wire_id - first_wire_id;
}
constexpr strict_signal_t strict(signal_t signal) {
  AOC_ASSERT(is_value(signal), "Cannot convert provided signal to value");
  return static_cast<strict_signal_t>(signal);
}

enum class operation_t {
  value_op,
  alias_op,
  not_op,
  and_op,
  or_op,
  lshift_op,
  rshift_op,
};

struct gate_t {
  operation_t op{operation_t::value_op};
  signal_t lhs{0};
  signal_t rhs{0};
};

// Include spaces to avoid trimming later
using namespace std::string_view_literals;
static constexpr const auto arrow_str = " -> "sv;
static constexpr const auto not_str = "NOT "sv;
// Don't include spaces here because we've already trimmed spaces
static constexpr const auto and_str = "AND"sv;
static constexpr const auto or_str = "OR"sv;
static constexpr const auto lshift_str = "LSHIFT"sv;
static constexpr const auto rshift_str = "RSHIFT"sv;

std::tuple<std::vector<gate_t>, signal_t, signal_t> parse(
    const std::string& filename) {
  auto name_to_id = aoc::name_to_id(first_wire_id);
  const auto get_signal = [&](std::string_view wire_name) -> signal_t {
    if (std::isdigit(wire_name[0])) {
      return aoc::to_number<signal_t>(wire_name);
    }
    return name_to_id.intern(wire_name);
  };
  std::vector<gate_t> inputs;
  for (std::string_view line : aoc::views::read_lines(filename)) {
    const auto [gate_str, wire_name] = aoc::split_once(line, arrow_str);
    const auto output_wire = get_signal(wire_name);
    gate_t gate{};
    if (gate_str.starts_with(not_str)) {
      gate.op = operation_t::not_op;
      gate.rhs = get_signal(gate_str.substr(not_str.size()));
    } else if (!gate_str.contains(' ')) {
      if (std::isdigit(gate_str[0])) {
        gate.op = operation_t::value_op;
      } else {
        gate.op = operation_t::alias_op;
      }
      gate.lhs = get_signal(gate_str);
    } else {
      const auto [lhs, op, rhs] = aoc::split_to_array<3>(gate_str, ' ');
      gate.lhs = get_signal(lhs);
      gate.op = [&]() {
        if (op == and_str) {
          return operation_t::and_op;
        }
        if (op == or_str) {
          return operation_t::or_op;
        }
        if (op == lshift_str) {
          return operation_t::lshift_op;
        }
        if (op == rshift_str) {
          return operation_t::rshift_op;
        }
        AOC_UNREACHABLE("Invalid operation!");
      }();
      gate.rhs = get_signal(rhs);
    }
    const auto index = input_id(output_wire);
    inputs.resize(std::max(static_cast<size_t>(index + 1), inputs.size()));
    inputs[index] = gate;
  }
  return {inputs, get_signal("a"), get_signal("b")};
}

constexpr strict_signal_t solve_wire(const std::span<gate_t> inputs,
                                     const signal_t start) {
  if (is_value(start)) {
    return strict(start);
  }
  auto& gate = inputs[input_id(start)];
  // Note that all operations would undergo integral promotion anyway,
  // so we just cast the result instead of the operands
  switch (gate.op) {
    using enum operation_t;
    case value_op:
      return strict(gate.lhs);
    case alias_op:
      gate.lhs = solve_wire(inputs, gate.lhs);
      return strict(gate.lhs);
    case not_op:
      gate.rhs = solve_wire(inputs, gate.rhs);
      // We need to explicitly mask the lower 16 bits to avoid UB
      return strict(~gate.rhs & (first_wire_id - 1));
    default:
      gate.lhs = solve_wire(inputs, gate.lhs);
      gate.rhs = solve_wire(inputs, gate.rhs);
      break;
  }
  switch (gate.op) {
    using enum operation_t;
    case and_op:
      return strict(gate.lhs & gate.rhs);
    case or_op:
      return strict(gate.lhs | gate.rhs);
    case lshift_op:
      return strict(gate.lhs << gate.rhs);
    case rshift_op:
      return strict(gate.lhs >> gate.rhs);
    default:
      AOC_UNREACHABLE("Operation should have been covered before!");
  }
}

template <bool override>
strict_signal_t solve_case(const std::string& filename) {
  auto [inputs, a, b] = parse(filename);
  strict_signal_t result = 0;
  {
    auto inputs_copy = [&]() {
      if constexpr (override) {
        return inputs;
      } else {
        return std::move(inputs);
      }
    }();
    result = solve_wire(inputs_copy, a);
  }
  if constexpr (override) {
    inputs[input_id(b)] = gate_t{
        .op = operation_t::value_op,
        .lhs = result,
        .rhs = 0,
    };
    result = solve_wire(inputs, a);
  }
  return result;
}

int main() {
  std::println("Part 1");
  AOC_EXPECT_RESULT(114, solve_case<false>("day07.example"));
  AOC_EXPECT_RESULT(16076, solve_case<false>("day07.input"));
  std::println("Part 2");
  AOC_EXPECT_RESULT(28, solve_case<true>("day07.example"));
  AOC_EXPECT_RESULT(2797, solve_case<true>("day07.input"));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2015/src/day07.rs
use aoc::string::NameToId;

// A signal can be a wire ID or an an actual value
// We unify them by storing values as 16-bit numbers,
// and the wire ID as numbers that are higher than all signal values
type SignalT = u32;
type WireIdT = SignalT;
type StrictSignalT = u16;
const FIRST_WIRE_ID: WireIdT = 1 << 16;
const fn is_value(signal: SignalT) -> bool {
    return signal < FIRST_WIRE_ID;
}
const fn input_id(wire_id: WireIdT) -> usize {
    assert!(wire_id >= FIRST_WIRE_ID, "Invalid wire ID");
    return (wire_id - FIRST_WIRE_ID) as usize;
}
const fn strict(signal: SignalT) -> StrictSignalT {
    assert!(is_value(signal), "Cannot convert provided signal to value");
    return signal as StrictSignalT;
}

#[derive(Clone)]
enum OperationT {
    ValueOp,
    AliasOp,
    NotOp,
    AndOp,
    OrOp,
    LShiftOp,
    RShiftOp,
}

#[derive(Clone)]
struct GateT {
    op: OperationT,
    lhs: SignalT,
    rhs: SignalT,
}

impl GateT {
    fn empty() -> GateT {
        GateT {
            op: OperationT::ValueOp,
            lhs: 0,
            rhs: 0,
        }
    }
}

// Include spaces to avoid trimming later
const ARROW_STR: &str = " -> ";
const NOT_STR: &str = "NOT ";
// Don't include spaces here because we've already trimmed spaces
const AND_STR: &str = "AND";
const OR_STR: &str = "OR";
const LSHIFT_STR: &str = "LSHIFT";
const RSHIFT_STR: &str = "RSHIFT";

fn parse(filename: &str) -> (Vec<GateT>, SignalT, SignalT) {
    let mut name_to_id = NameToId::from(FIRST_WIRE_ID as usize);
    let mut get_signal = |wire_name: &str| -> SignalT {
        if wire_name.chars().next().unwrap().is_digit(10) {
            return wire_name.parse().unwrap();
        }
        return name_to_id.intern(wire_name) as u32;
    };
    let mut inputs = Vec::<GateT>::new();
    for line in std::fs::read_to_string(filename).unwrap().lines() {
        let (gate_str, wire_name) = line.split_once(ARROW_STR).unwrap();
        let output_wire = get_signal(wire_name);
        let mut gate = GateT::empty();
        if gate_str.starts_with(NOT_STR) {
            gate.op = OperationT::NotOp;
            gate.rhs = get_signal(&gate_str[NOT_STR.len()..]);
        } else if !gate_str.contains(' ') {
            if gate_str.chars().next().unwrap().is_digit(10) {
                gate.op = OperationT::ValueOp;
            } else {
                gate.op = OperationT::AliasOp;
            }
            gate.lhs = get_signal(gate_str);
        } else {
            let parts = gate_str.split(' ').collect::<Vec<_>>();
            let [lhs, op, rhs] = parts.as_slice() else {
                panic!("Bad parsing");
            };
            gate.lhs = get_signal(lhs);
            gate.op = if *op == AND_STR {
                OperationT::AndOp
            } else if *op == OR_STR {
                OperationT::OrOp
            } else if *op == LSHIFT_STR {
                OperationT::LShiftOp
            } else if *op == RSHIFT_STR {
                OperationT::RShiftOp
            } else {
                unreachable!("Invalid operation!");
            };
            gate.rhs = get_signal(rhs);
        }

        let index = input_id(output_wire);
        inputs.resize(
            std::cmp::max(index + 1 as usize, inputs.len()),
            GateT::empty(),
        );
        inputs[index] = gate;
    }
    (inputs, get_signal("a"), get_signal("b"))
}

fn solve_wire(inputs: &mut [GateT], start: SignalT) -> StrictSignalT {
    if is_value(start) {
        return strict(start);
    }
    let index = input_id(start);
    let gate = inputs[index].clone();
    use OperationT::*;
    match gate.op {
        ValueOp => {
            return strict(gate.lhs);
        }
        AliasOp => {
            inputs[index].lhs = solve_wire(inputs, gate.lhs) as SignalT;
            return strict(inputs[index].lhs);
        }
        NotOp => {
            inputs[index].rhs = solve_wire(inputs, gate.rhs) as SignalT;
            return !strict(inputs[index].rhs);
        }
        _ => {
            inputs[index].lhs = solve_wire(inputs, gate.lhs) as SignalT;
            inputs[index].rhs = solve_wire(inputs, gate.rhs) as SignalT;
        }
    };
    let gate = &inputs[index];
    match gate.op {
        AndOp => {
            return strict(gate.lhs) & strict(gate.rhs);
        }
        OrOp => {
            return strict(gate.lhs) | strict(gate.rhs);
        }
        LShiftOp => {
            return strict(gate.lhs) << strict(gate.rhs);
        }
        RShiftOp => {
            return strict(gate.lhs) >> strict(gate.rhs);
        }
        _ => {
            unreachable!("Operation should have been covered before!");
        }
    }
}

fn solve_case1(filename: &str) -> StrictSignalT {
    let (mut inputs, a, _) = parse(filename);
    solve_wire(&mut inputs, a)
}

fn solve_case2(filename: &str) -> StrictSignalT {
    let (mut inputs, a, b) = parse(filename);
    let mut inputs_copy = inputs.clone();
    let result = solve_wire(&mut inputs_copy, a);
    inputs[input_id(b)] = GateT {
        op: OperationT::ValueOp,
        lhs: result as SignalT,
        rhs: 0,
    };
    solve_wire(&mut inputs, a)
}

fn main() {
    println!("Part 1");
    assert_eq!(114, solve_case1("day07.example"));
    assert_eq!(16076, solve_case1("day07.input"));
    println!("Part 2");
    assert_eq!(28, solve_case2("day07.example"));
    assert_eq!(2797, solve_case2("day07.input"));
}
